src/acl/external/time_quota/ext_time_quota_acl.cc

```diff
@@ -76,15 +76,6 @@ static int pauseLength = 300;
 static FILE *logfile = stderr;
 static int tq_debug_enabled = false;
 
-static void open_log(const char *logfilename)
-{
-    logfile = fopen(logfilename, "a");
```

Comment:
> PR description: Debug logging to own file does not work well when there are
multiple helpers, which could clobber each other's output.

I have re-checked what fopen("a") means. The above assertion is rather misleading because fopen("a") guarantees no clobbering by multiple processes writing debugging messages as long as each underlying write(2) call is given the whole message (i.e. no partial write(2)s due to fwrite() buffering). If all actual debugging messages in this helper are smaller than fwrite() buffers, then there will be no clobbering. I speculate that they are smaller in supported environments.

> It's better to log through squid and find the output in cache.log.

Squid's cache.log is shared among multiple SMP kids. Its code uses exactly the same fopen("a") approach. Thus, if helpers do have clobbering problems in some environments, then migrating to cache.log will make those clobbering problems _worse_ in SMP Squid cases!

I suggest dropping the clobbering claim and appealing to debugs() upgrade overheads instead. I edited PR description accordingly. Please review and undo or adjust as needed!

P.S. I have not yet fully checked what it would take to solve the commSetCloseOnExec() problem. I am currently relying exclusively on your assessment that there is no trivial solution there.

Reply:
> fopen("a") guarantees no clobbering by multiple processes writing debugging messages as long as each underlying write(2)

The MacOS manual page reads:
```
     “a”     Open for writing.  The stream is positioned at the end of the file.
             Subsequent writes to the file will always end up at the then
             current end of file, irrespective of any intervening fseek(3) or
             similar.  Create the file if it does not exist.
```
In other words, it guarantees that each write happens at the end of file, but I can't find any mention of the guarantee you mention.
In a worst-case scenario:

1. we have no guarantee that printf() will issue one call to write() per call (although it's likely, but probably implementation-dependent)
2. we have no guarantee that a call to write() will not be interrupted

Let's imagine a scenario where we have two processes writing to the same file in "a" mode.
* A issues its command, writes some output but to complete the write the OS requires allocating a block on storage. A gets suspended while the OS does that
* After the OS allocates the storage but before A gets a chance to run again, B gets a chance to run. B has a pending write, and issues it, just in the middle of A's output
* A now gets a chance to finish its write operation. It does so at the end of file, without overwriting what B has just written to disk, but now the last two entries written to the file are corrupted.

Probably not a big deal, but a possible scenario.

Squid suffers from the same problem, although probably to a lesser extent: I expect that a typical installation will have many more helpers than squid worker processes running, as most helpers do not support multithreaded operations and more processes are needed to match squid's parallelism

Reply:
Are you _sure_ we should be spending time on this? If you are, I will find the time to clarify and probably invalidate the proposed scenario. As I mentioned earlier, I recall being satisfied when I studied the issue at depth several years ago. My current recollection may be faulty, of course.

> Squid suffers from the same problem, although probably to a lesser extent

Modern busy SMP Squids in ALL,9 debugging mode would probably be much, much more susceptible to this problem if it were real. FWIW, I have never seen them clobber their (often large) debugs() messages. I could miss some rare cases, of course. However, my prior research was not based on these observations. It was based on standards and their implementations.

Reply:
If you want a bit more information for making that "spend more time on this" decision, consider reading [Linus](https://lkml.iu.edu/hypermail/linux/kernel/0208.0/0251.html) and [others](https://stackoverflow.com/questions/12942915/understanding-concurrent-file-writes-from-multiple-processes#comment94938866_12943431) (while carefully filtering out noise and irrelevant details).

Reply:
Got it.
If one were to pick nits, `DebugChannel::writeToStream` uses fprintf(3),
not write(2). Given the short size of the output, it probably doesn't make a difference. I wonder if there would be benefit by using writev() instead.

Reply:
> If one were to pick nits,

... then one should not ignore the word "underlying" in "underlying write(2)" in the [original statement](https://github.com/squid-cache/squid/pull/1872#discussion_r1690275970).

> Given the short size of the output, it probably doesn't make a difference.

Yes, in typical environments (at least), stdio buffers are evidently large enough for typical cache.log messages. I may have even checked what the defaults are and found them satisfactory when I was researching this, but I am not sure.

> I wonder if there would be benefit by using writev(2) instead.

Most likely, we want to get rid of the remaining printf() and such in debugging stream implementation. Once that is done, we will probably just write(2) a single buffer (filled by in-memory std::ostream writes a la ToSBuf()) -- no need to build a vector for writev(2).